monteCarlo and main

Jenna993 opened the pull request:
has an error during compilation about 'print'. I believe it is due to not having compiled versions of board and node.

Jenna993 committed:
monteCarlo and main

has an error during compilation about print. I believe it is due to not having compiled versions of board and node.

emanguy commented:
Cool I'll check it later

emanguy commented:
I'm going to bed now but before I actually review this the reason it's giving you an error about "print" is because the function is `printf()` not `print()`. Try updating that and making another commit.

Jenna993 committed:
fixed code

fixed print issue. I thought I had printf there but I somehow changed it sometime in the process.

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+	int timeLimit = 3000;		//the time limit to be given to the function
+	int winner;					//holds the value for winner
+	struct board * gameBoard;	//gameboard
+	initBoard(gameBoard);		//initialize the gameBoard
```
Don't make a `struct board` pointer here, just declare a normal struct and pass `initBoard()` a reference to gameBoard.

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+	while(gameContinues > 0)
+	{
+		//picks the move
+		pickMove(gameBoard,nodeLimit,timeLimit, boardMove);
```
You'll either need to make boardMove an array or initialize it with `malloc()`, as you're not currently initializing it. `pickMove()` should just modify the passed array.

emanguy replied:
Going along with the comment on line 290, be sure to make gameBoard a normal struct and just pass a reference to it in this function.

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+		{
+
+			//Check if other player can move
+			flipMove(gameBoard->whoseMove);
```
`flipMove()` is a function local to board.c and should not be accessed as it is not in the header file. This can easily be replaced with a ternary if statement.

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+
+			//Check if other player can move
+			flipMove(gameBoard->whoseMove);
+			pickMove(gameBoard,nodeLimit,timeLimit,boardMove);
```
Pass a reference to gameBoard

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+			//Continue the game
+			else
+			{
+				placePiece(gameBoard,boardMove);
```
Pass a reference to gameBoard

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+		}
+		else
+		{
+			placePiece(gameBoard, boardMove);
```
Pass a reference to gameBoard

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+			if(boardMove[0] == -1 && boardMove[1] == -1)
+			{
+				gameContinues =-10;
+				winner = determineWinner(gameBoard);
```
Pass a reference to gameBoard

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+#include <stdlib.h>
+#include <stdio.h> 
+#include "board.h"
+#include "math.h"
```
This isn't a file local to the repository, did you forget to add it or did you mean to put this in angle brackets?

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+	node *temp;					//temporary pointer for node
+	node *child;				//child node pointer
+
+	struct board * boardState;  //gameboard pointer for use in simulation    
```
This doesn't need to be a pointer

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+
+void pickMove(struct board* gameBoard, int nodeLimit, int timeLimit, int* move)
+{
+	node *root; 				//root node
```
This doesn't need to be a pointer

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+
+	int * allActions;           //list of all actions returned by getMoves
+	int * action;				//action to be used for expansion
+	int * boardMove;            // boardMove for simulation
```
You either need to create these three arrays with `malloc()` or you need to make them normal arrays.

Jenna993 replied:
if I make these normal arrays what size do I need to make them?

emanguy replied:
Let's go with 200 for allmoves

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+		// {
+		// 	move[0] = -1;
+		// 	move[1] = -1;
+		// 	flag == -100;
```
Did you mean for this to be commented? I see that you trip the flag on line 241 but I think that only happens when you can't expand the root node.

Jenna993 replied:
I commented it out when I was unsure if I needed it there. I reorganized my code so that I did not need to use the flag there, but I left it there just in case

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+	clock_t startTime, timeCheck; //used to calculate the time that has passed
+
+	startTime = clock();				//track time limit
+	initializeRoot(root, gameBoard);	  // initialize root
```
When you turn root and gameBoard into normal structs, pass these parameters as references to the structs

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+
+
+		
+		getMoves(gameBoard, allActions);
```
You should be getting the moves from the **board** property of the selected node.

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+		if(numOfActions != 0)
+		{
+
+			for(i = 0; i < numOfActions; i+2)
```
Should be i += 2, not i+2

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+		//3. Simulate the game board on the new node until the game is complete 
+		for(j = 0; j < selected->numChildren; j++)
+		{
+			boardState = selected->board;
```
If you want the board state in a variable, you should use `copyBoard()`

emanguy replied:
Also you should be using the child's board that you just expanded to. `selected->children[j]->board`

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+				{
+					//Randomly select a move to make
+					randomValue = numOfActions / 2;
+					randomValue = rand() % randomValue + 1;  //may not be plus 1
```
You're right, it's not plus one.

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+					randomValue = numOfActions / 2;
+					randomValue = rand() % randomValue + 1;  //may not be plus 1
+					boardMove[0] = allActions[randomValue];
+					boardMove[1] = allActions[randomValue + 1]; 
```
These two lines should be `allActions[randomValue * 2]` and `allActions[randomValue * 2 + 1]`

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+				{
+
+					//Check if other player can move
+					flipMove(boardState->whoseMove);
```
You can't use this, it's not in board.h. Use a ternary if statement instead.

emanguy replied:
Note: if you really want me to give you access to the `flipMove()` function you're gonna need to pass a reference to whoseMove instead of passing by value. Let me know if you need me to expose `flipMove()`.

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+
+					//Check if other player can move
+					flipMove(boardState->whoseMove);
+					pickMove(boardState, nodeLimit, timeLimit, boardMove);
```
No need for recursion, this would probably make you go on forever. Just do the same thing you did in lines 128-152.

Jenna993 replied:
I did not mean to put that there, lol. I thought I deleted that... whoops

emanguy replied:
Lol yep, this is why we do code reviews

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+						winner = determineWinner(boardState);
+						if(winner == 0)
+						{
+							selected->numSimulations += 1;
```
`selected->children[j]->numSimulations`

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+						}
+						else if(winner == 1)
+						{
+							if(boardState->whoseMove == 1)
```
You should be using the expanded child's board here. `selected->children[j]->board->whoseMove`

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+						{
+							if(boardState->whoseMove == 1)
+							{
+								selected->numWins += 1;
```
`selected->children[j]->numWins`

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+							{
+								selected->numWins += 1;
+							}
+							selected->numSimulations += 1;
```
`selected->children[j]->numSimulations`

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+						}	
+						else if(winner == 2)
+						{
+							if(boardState->whoseMove == 2)
```
`selected->children[j]->board->whoseMove`

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+						{
+							if(boardState->whoseMove == 2)
+							{
+								selected->numWins += 1;
```
`selected->children[j]->numWins`

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+							{
+								selected->numWins += 1;
+							}
+							selected->numSimulations += 1;
```
`selected->children[j]->numSimulations`

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+		 	sum = 0;
+		 	numWins = 0;
+			numSims = 0;	
+		 	parentNode = newNode->parent;
```
You'll need to sum up the wins/simulations on the newly expanded child nodes of the selected node first, so hold off on this for now.

Jenna993 replied:
Does this comment have to do with the comment 2 down from here? I think that I just need to move this outside the while loop right?

emanguy replied:
Yeah. The comment immediately below that one should give you some direction for the back propagation

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+			numSims = 0;	
+		 	parentNode = newNode->parent;
+		 	//make sure that the parentNode has children
+		 	if(parentNode->numChildren > 0)
```
If you just moved up to the parent node from a child node, it's guaranteed to have children right?

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+		 	//make sure that the parentNode has children
+		 	if(parentNode->numChildren > 0)
+		 	{
+		 		for(i = 0; i<parentNode->numChildren; i++)
```
I think you want to operate on the children of the selected node, not the selected node's parent.

emanguy replied:
I think you should put this for loop outside the while loop. Calculate sum (the total wins from the newly expanded child nodes) and the number of total simulations among the child nodes and add those values to parent nodes as you ascend the tree.

Jenna993 replied:
So I should move my loop outside of the while loop and leave the code alone?

emanguy replied:
The loop where you're summing the expanded children should be outside the back propagation loop. Sum the children first, then add that number to every parent node.

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+		 		sum = numWins + parentNode->numWins;
+		 		parentNode->numWins = sum;
+		 		sum = numSims + parentNode->numSimulations;
+		 		parentNode->numSimulations = sum;
```
Once you have the sum of the number of wins and number of simulations from the newly expanded child nodes, you should just plus-equals their sum on every node as you go up the tree. Take this out of the for loop and leave it in the while loop.

```c
// Do this to every parent node as you ascend the tree
parentNode->numWins += sum; // Sum of the total wins resulting from the simulations
parentNode->numSimulations += totalSimulations; // Or whatever variable you use to store the total number of simulations you just did
parentNode = parentNode->parent;
```

Jenna993 replied:
The code on lines 223 - 226 are not in the for loop... The for loop adds up the children's numWins and numSimulations and then outside of the for loop that value is added to the parentNodes.

emanguy replied:
The while loop outside this for loop makes every parent sum its children even if they weren't just expanded. There should only be one level of looping.

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+    //check that there were moves to be made
+    if(root->numChildren != 0)
+    {
+  		//loops to find the node that was most visited
```
You want to return the immediate child node with the highest UCT value, not the most frequently visited immediate child node.

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+		move[0] = -1; //root has no children nodes which means there is no moves to be made
+		move[1] = -1; 	
+	}
+  	deconstructTree(root);
```
When root is a normal struct, pass `deconstructTree()` a reference to it.

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,334 @@
+			child = root->children[i];
+			numSims = child->numSimulations;
+			//finds the most visited node to choose the move to make
+		 	if(numSims > bestValue)
```
So calculate UCT of the child node first, then figure out if it's the one with the highest UCT value.

emanguy reviewed (changes requested):
There's a lot of little things that need to be fixed, most importantly you're making pointers without assigning any data to them. Fix the issues I commented on and I'll review again.

Jenna993 committed:
fixed issues

I had to remove one of the typedef's in node.h  to compile. It compiles and runs for a little while. The problem is there is a segfault like 3 cycles in. For some reason the number of children that root has is never incremented. To get it to run past 1 loop I had to comment out my exit if statement so that it did not immediately exit after one loop. There may be something not quite right with deconstruct tree also because it gave a message about freeing memory. I see debug statements from node.c saying that there were children nodes created during expansion. I am not sure what is going on. It is either an issue in my code or node.c

Jenna993 commented:
On line 84 of my most recent commit I added a line that I may not have needed to add. Just let me know if I need to remove it and resubmit. I was not sure if I needed to change the value of newNode as the program loops or not.

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,383 @@
+	node parentNode;           //node to hold the parentNode for backPropagation
+	node newNode;              // tempNode for looping
+	node selected;				//selected value for expansion
+	node child;				//child node pointer
```
I only wanted the root node to not be a pointer, not the rest of this.

Every child node under the root is dynamically allocated so it's more convenient to have a statically made root node while the children (pointers) are created with `expand()`.

emanguy replied:
Actually, I take it back. You should make all of these pointers and just malloc() the root node. That way you can just call deconstructTree() on root without having to worry about accidentally calling free() on it.

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,383 @@
+			//check if the uct value for each child is better than the one before
+			if(uctValue > bestValue)
+			{
+				selected = child;
```
Make sure "selected" is a pointer so we're operating on the actual tree nodes and not copies of the nodes.

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,383 @@
+		newNode = selected;
+		printf("After selection\n");
+
+		getMoves(&selected.board, allActions);
```
When "selected" is a pointer, make sure this is `&selected->board`

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,383 @@
+				action[0] = allActions[i];
+				action[1] = allActions[i+ 1];
+				//You need to implement a choose move here and then send that move as action.
+				expand(&selected, action);
```
Since selected will be a pointer, this will need to be `expand(selected, action)`

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,383 @@
+			// Do this to every parent node as you ascend the tree
+			parentNode.numWins += numWins; // Sum of the total wins resulting from the simulations
+			parentNode.numSimulations += numSims; // Or whatever variable you use to store the total number of simulations you just did
+			parentNode = *parentNode.parent;
```
You'll need to update this code to reflect parentNode being a pointer.

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,383 @@
+		}
+		parentNode = selected;
+		
+		while(parentNode.parent !=NULL)
```
I think this may cause the root node to not be updated. Change the condition to `parentNode != NULL`.

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,383 @@
+   		printf("root does not equal zero\n");
+
+		//UCT calculation to select next node 
+		for(i = 0; i < root.numChildren; i++)
```
Don't forget to update to reflect the struct pointer change in this loop

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,383 @@
+
+		printf("Selection\n");
+		//UCT calculation to select next node 
+		for(i = 0; i < newNode.numChildren; i++)
```
This for loop never traverses down the tree. Make sure you're not just selecting nodes immediately under the root by doing this for loop for every "selected" node while the "selected" node has children.

emanguy replied:
I.E. put it in a while loop that goes until you select a node with no children.

emanguy commented on monteCarloTreeSearch.c:
```diff
@@ -0,0 +1,383 @@
+	}
+	printf("deconstruct tree\n");
+
+  	deconstructTree(&root);
```
Remember to update for the node pointer changes.

emanguy reviewed (changes requested):
Looking much better. Just have to make selection move down the tree (you don't want to expand an already expanded node!) and make the node struct pointer changes.

emanguy committed:
Merge branch 'master' into othello

emanguy committed:
Update makefile

Jenna993 committed:
added while loop and pointers

made changes

Jenna993 committed:
Merge remote-tracking branch 'origin/othello' into othello

Jenna993 committed:
selected = root change

emanguy committed:
Get that DS Store outta here

emanguy committed:
Merge branch 'othello' of github.com:emanguy/cs483-mcts-othello into othello

emanguy committed:
Condense if statement

emanguy committed:
It works maybe

Jenna993 committed:
fixed typdef

typedef fixed so no error is received

Jenna993 committed:
Merge remote-tracking branch 'origin/othello' into othello

Jenna993 committed:
removed debug statements

removed debug statements, changed nodeLimit, printed the move to be made

emanguy committed:
Add single threaded playback output and playback program
